hera_cal/abscal.py

```diff
@@ -3693,6 +3694,158 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+            hdm.select(antenna_nums=all_data_ants,
+                       keep_all_metadata=False)
+        hdm.inflate_by_redundancy()
+        # also make sure to only include baselines present in hdd.
+        hdm.select(bls=hdd.bls)
+
+    _, model_flags, _ = hdm.build_datacontainers()
+
+    data_flags = synthesize_ant_flags(data_flags, threshold=ant_threshold)
+    model_flags = synthesize_ant_flags(model_flags, threshold=ant_threshold)
+
+    lst_center = hdd.lsts[hdd.Ntimes // 2] * 12 / np.pi
+    field_str = 'LST={lst_center:%.2f} hrs'
+
+    if refant is None:
+        # dummy refant. Could be completely flagged for all we know.
+        refant_init = str(hdd.ant_1_array[0])
+    else:
+        refant_init = str(refant)
+    # initialize HERACal to store new cal solutions
+    hc = UVCal()
+    hc = hc.initialize_from_uvdata(uvdata=hdd, gain_convention='divide', cal_style='sky',
+                                   ref_antenna_name=refant_init, sky_catalog=f'{model_file}',
+                                   metadata_only=False, sky_field=field_str, cal_type='gain')
+    hc.write_calfits(output_filename, clobber=clobber)
+    hc = io.HERACal(output_filename)
+    hc.read()
+    hc.update(flags=data_flags)
+    # generate cal object from model to hold model flags.
+    hcm = UVCal()
+    hcm = hcm.initialize_from_uvdata(uvdata=hdm, gain_convention='divide', cal_style='sky',
+                                     ref_antenna_name=refant_init, sky_catalog=f'{model_file}',
+                                     metadata_only=False, sky_field=field_str, cal_type='gain')
+    hcm.write_calfits(output_filename + '.tmp')
+    hcm = io.HERACal(output_filename + '.tmp')
+    hcm.read()
+    hcm.update(flags=model_flags)
+    # init all gains to unity.
+    hcm.gain_array[:] = 1. + 0.j
+    hc.gain_array[:] = 1. + 0.j
+    hc.flag_array = hc.flag_array | hcm.flag_array
+
+    abscal_gains, abscal_flags, _, _ = hc.build_calcontainers()
+
+    # load in weights if supplied.
+    if auto_file is not None:
+        hda = io.HERAData(auto_file)
+        bls = [ap for ap in hda.get_antpairs() if ap[0] == ap[1]]
+        auto_data, auto_flags, auto_nsamples = hda.read(bls=bls)
+        data, data_flags, data_nsamples = hdd.read()
+        model, model_flags, model_nsamples = hdm.read()
+        # generate wgts from autocorrelations
+        wgts = build_data_wgts(data_flags=data_flags, data_nsamples=data_nsamples, model_flags=model_flags,
+                               autocorrs=auto_data, auto_flags=auto_flags)
+    else:
+        wgts = None
+
+    # select refant if not specified.
+    if refant is None:
+        refant = pick_reference_antenna(abscal_gains, abscal_flags, hc.freqs, per_pol=True)
+    hc.ref_antenna_name = str(refant)
+    hcm.ref_antenna_name = str(refant)
+
+    for iter in range(iterations):
+        abscal = AbsCal(data=hdd, model=hdm, wgts=wgts)
```

Comment:
you shouldn't have to reinstantiate the abscal class here. you can simply apply the calibration to the data attached to abscal object

Reply:
I've rewritten this to update the AbsCal data dictionaries.